Fix Tar timestamp conversion from/to string and DateTimeOffset

carlossanlop opened the pull request:
The tests that made comparisons between expected timestamps and timestamps extracted from extended attributes (string -> double -> DateTimeOffset) were intermittently failing due to a precision bug. The main suspicion is that double is losing precision, so I was given the suggestion [here](https://github.com/dotnet/runtime/pull/70325#discussion_r902041685) to change it to decimal. In that same comment, I added more decimals when parsing (from 6 to 9) to help with the precision as well.

Fixes https://github.com/dotnet/runtime/issues/69474
Fixes https://github.com/dotnet/runtime/issues/70060

Re-enable test disabled here: https://github.com/dotnet/runtime/pull/69997

carlossanlop committed:
Change double to decimal in timestamp conversions to preserve precision

carlossanlop committed:
Re-enable disabled test

carlossanlop committed:
Adjust assert message

carlossanlop commented:
/azp run runtime-extra-platforms

danmoseley commented:
Do we understand where the loss occurs? It seems unlikely it's long-double-long since as @tannergooding  pointed out that would require a value over 2^52. I wonder whether we are papering over an issue that may manifest elsewhere.

If we are lacking a local repro to debug, we could disable the test for now (or make it tolerant), and leave the issue open. Then later add logging/asserts and loop it until we find a debuggable repro?

eerhardt commented:
> It seems unlikely it's long-double-long since as @tannergooding pointed out that would require a value over 2^52.

https://github.com/dotnet/runtime/blob/7ab7f83fb1bdfed781ceb8da066986184ac782a1/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs#L122-L123

Today (GMT: Tuesday, June 21, 2022 2:39:51 PM), there have been 63,823,041,591 seconds since 1/1/1970.

In the above code, we are multiplying seconds since epoch by `TimeSpan.TicksPerSecond` (10,000,000):

![image](https://user-images.githubusercontent.com/8291187/174828236-84de9881-b3f4-4530-ab6c-38ceb4bae3df.png)

eerhardt commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -143,8 +143,8 @@ internal static bool TryGetDateTimeOffsetFromTimestampString(Dictionary<string,
-            double secondsSinceEpoch = GetSecondsSinceEpochFromDateTimeOffset(timestamp);
-            return secondsSinceEpoch.ToString("F9", CultureInfo.InvariantCulture); // 6 decimals, no commas
+            decimal secondsSinceEpoch = GetSecondsSinceEpochFromDateTimeOffset(timestamp);
+            return secondsSinceEpoch.ToString("F9", CultureInfo.InvariantCulture); // 9 decimals, no commas
```
Don't we want `G9`?

![image](https://user-images.githubusercontent.com/8291187/174835319-00f5ab2c-c8b7-4333-a612-abd16508d34b.png)

eerhardt replied:
Hmm, maybe `9` is too small.

![image](https://user-images.githubusercontent.com/8291187/174836196-8bd4b907-0740-4f24-9f4d-927495aa815f.png)

eerhardt replied:
Maybe we just want `G`...

https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings#general-format-specifier-g

tannergooding replied:
Just `G` (the default) will print the shortest roundtrippable string. `G9` will print the shortest string or up to `9` significant digits, whichever is lesser. However, it will also print exponentials.

There is actually a decent amount of "inconsistency" in how the formatting APIs work between the different format specifiers and its a bit frustrating at times, but we can't really change existing ones due to back-compat.

carlossanlop replied:
> However, it will also print exponentials.

I had initially tested `G` and because it shows `E+XX` like in the screenshot above, I decided to use `F` instead.

@tannergooding just to make sure we are all in agreement, using `F` is ok?

eerhardt replied:
It won't print `E+XX` if you just use `G`

```C#
return secondsSinceEpoch.ToString("G", CultureInfo.InvariantCulture);
```

tannergooding replied:
> It won't print E+XX if you just use G

It can. `G` picks the more compact of `F` or `E`, but without including trailing non-significant digits.

tannergooding replied:
Gah, nevermind. This is a case where just `G` for `decimal` always picks `F` (basically, and therefore differentiating from the docs/convention), but `G##` picks `F` or `E` an.

I refer back to my earlier statement
> There is actually a decent amount of "inconsistency"

tannergooding replied:
> In what scenarios?

That table is quite a bit out of date and doesn't exactly line up with what the implementations do. `Double` and `Single` are just incorrect here.

eerhardt replied:
> There is actually a decent amount of "inconsistency"

Totally agree. I have to look it up every time.

> That table is quite a bit out of date and doesn't exactly line up with what the implementations do. Double and Single are just incorrect here.

😢 . Can we get it updated? I think people really rely on this kind of information (like I am above).

eerhardt replied:
https://github.com/dotnet/docs/issues/29952 has the full details and repro steps.

carlossanlop replied:
> In what scenarios?

```cs

double d = 1655815010.19999999998;
Console.WriteLine(d.ToString("G9", CultureInfo.InvariantCulture)); // 1.65581501E+09
d = 1.19999999998;
Console.WriteLine(d.ToString("G9", CultureInfo.InvariantCulture)); // 1.2
d = 123456789.19999999998;
Console.WriteLine(d.ToString("G9", CultureInfo.InvariantCulture)); // 123456789   (lol wat?)
```

eerhardt replied:
We are using a `decimal` in the code.

carlossanlop replied:
With decimal, `F9` works fine:

```cs
decimal d = 1655815010.199999998M;
Console.WriteLine(d.ToString("F9", CultureInfo.InvariantCulture)); // 1655815010.199999998
d = 1.199999998M;
Console.WriteLine(d.ToString("F9", CultureInfo.InvariantCulture)); // 1.199999998
d = 123456789.199999998M;
Console.WriteLine(d.ToString("F9", CultureInfo.InvariantCulture)); // 123456789.199999998
```

But `G9` behaves similarly to the example above for double:

```cs
decimal d = 1655815010.199999998M;
Console.WriteLine(d.ToString("G9", CultureInfo.InvariantCulture)); //1.65581501E+09
d = 1.199999998M;
Console.WriteLine(d.ToString("G9", CultureInfo.InvariantCulture)); // 1.2
d = 123456789.199999998M;
Console.WriteLine(d.ToString("G9", CultureInfo.InvariantCulture)); // 123456789
```

carlossanlop replied:
I'll use `G` as suggested above, and as explained via chat.

I am also going to remove some calls to this conversion code where it's not necessary. This code to convert timestamps is only required when converting from gnu to pax or viceversa (to store atime and ctime as required by the format).

eerhardt commented on src/libraries/System.Formats.Tar/tests/TarEntry/TarEntry.Conversion.Tests.Base.cs:
```diff
@@ -128,13 +128,13 @@ private TarEntry ConvertAndVerifyEntry(TarEntry originalEntry, TarEntryType entr
                 {
-                    CompareDateTimeOffsets(initialNow, actualAccessTime);
-                    CompareDateTimeOffsets(initialNow, actualChangeTime);
+                    AssertExtensions.GreaterThanOrEqualTo(actualAccessTime, initialNow);
```
Why is this `GreaterThanOrEqualTo` and not just `Assert.Equal`?

carlossanlop replied:
In these cases, the `initialNow` timestamp is `DateTimeOffset.UtcNow`, which is generated shortly before invoking the constructors that take the `other` entry. Since the constructors generate the `mtime` timestamp automatically using `UtcNow`, I can't know the _exact expected_ value, but I can at least verify that the value is _larger than_ the timestamp I saved before calling the constructors.

The cases that do an `Equal` comparison are the ones where the _existing_ `mtime` is used as the value to store for `atime` and `ctime`, so I know the exact expected value to compare.

carlossanlop committed:
Reuse FileSystemInfo Last*TimeUtc fields, use implicit cast operator to store as DateTimeOffset.

carlossanlop committed:
Fix using wrong fieldname for ctime.

carlossanlop committed:
Using 'G' for decimal to string conversion. Adjust test to not require a dot. The DateTimeOffset comparison done afterwards should suffice.

carlossanlop commented:
@eerhardt the latest 3 commits take care of the following:
- Avoid using the conversion methods when it is not needed: There is a case where we create a `FileSystemInfo` object, and we can reuse 2 of the 3 `Last*TimeUtc` properties.
- Fixed a bug checking two times for a key in the dictionary, instead of two different ones.
- Using 'G' for string formatting of the decimal, as recommended. I also modified the related test to not require finding a dot character in the string. The important thing is that later that timestamp should be able to do the roundtrip, and the tests should pass.
  - The particular case where the string timestamp might not contain a dot and decimal portion is when the a v7/ustar/gnu entry is converted to pax, and the mtime field is then copied into the extended attributes dictionary. The mtime field in that case did not have a decimal portion, it was obtained directly from the old header mtime field, which holds an integer.

carlossanlop commented:
/azp run runtime-extra-platforms

carlossanlop committed:
Ensure timestamps are converted to long, not int.

carlossanlop commented:
/azp run runtime-extra-platforms

eerhardt commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs:
```diff
@@ -383,10 +383,10 @@ private void ReadPosixAndGnuSharedAttributes(Span<byte> buffer)
         {
             // Convert byte arrays
-            int aTime = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.ATime, FieldLengths.ATime));
+            long aTime = TarHelpers.GetTenBaseLongFromOctalAsciiChars(buffer.Slice(FieldLocations.ATime, FieldLengths.ATime));
```
It would be good to create a test that has a date time in 2039 to make sure we can handle dates past 2038.

carlossanlop replied:
Added tests for the epochalypse and the max upper limit in octal.

eerhardt commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -119,20 +119,20 @@ internal static bool IsAllNullBytes(Span<byte> buffer)
         // Converts the specified DateTimeOffset to the number of seconds that have passed since the Unix Epoch.
-        internal static double GetSecondsSinceEpochFromDateTimeOffset(DateTimeOffset dateTimeOffset) =>
-            ((double)(dateTimeOffset.UtcDateTime - DateTime.UnixEpoch).Ticks) / TimeSpan.TicksPerSecond;
+        internal static decimal GetSecondsSinceEpochFromDateTimeOffset(DateTimeOffset dateTimeOffset) =>
```
Does this need to be `internal`? Is it only called from this class?

carlossanlop replied:
Fixed

eerhardt commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs:
```diff
@@ -179,6 +181,14 @@ internal static int GetTenBaseNumberFromOctalAsciiChars(Span<byte> buffer)
+        // Converts the array to an octal base number, then transforms it to ten base and returns it.
+        internal static long GetTenBaseLongFromOctalAsciiChars(Span<byte> buffer)
+        {
+            string str = GetTrimmedAsciiString(buffer);
```
Do we really need to create an intermediate string just to parse it into an integer?

eerhardt replied:
I don't see a Utf8Parser API that parses the integer from an "octal" string. I see it supports `x` - hexstring.

carlossanlop replied:
I agree that a new allocation is not needed. This code can be improved to instead of returning a string, it returns an int representing the length of the ROS for slicing.

The method checks that the last character(s) in the ROS are either a `0` (null char) or a `32` (space). All other characters are not trimmed. Which means that if an unexpected non-numeric character is found, it will cause the conversion to fail.

Do you mind if I address this request later? I'd like to get this PR merged just for the DateTimeOffsets.

eerhardt replied:
> Do you mind if I address this request later?

I think that is fine.

eerhardt commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.Unix.cs:
```diff
@@ -58,8 +58,8 @@ public sealed partial class TarWriter : IDisposable
-            entry._header._aTime = TarHelpers.GetDateTimeOffsetFromSecondsSinceEpoch(status.ATime);
+            entry._header._mTime = info.LastWriteTimeUtc;
+            entry._header._aTime = info.LastAccessTimeUtc;
             entry._header._cTime = TarHelpers.GetDateTimeOffsetFromSecondsSinceEpoch(status.CTime);
```
From looking at the FileStatus code:

[https://github.com/dotnet/runtime/blob/fdc3b51da075066a5da489be2c169be85dbc117f/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs#L309-L318](https://github.com/dotnet/runtime/blob/fdc3b51da075066a5da489be2c169be85dbc117f/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs#L309-L318)

It looks like `status.XTime` gets added to `status.XTimeNsec` to make this DateTimeOffset. We will be respecting that for M and A time, but it looks like we don't respect that for C time now.

carlossanlop replied:
Good point. Changing this logic to:

```cs
            entry._header._cTime = DateTimeOffset.FromUnixTimeSeconds(status.CTime).AddTicks(status.CTimeNsec / 100 /* nanoseconds per tick */);
```

carlossanlop committed:
Add Epochalypse and Past-octal-limit timestamp tests

carlossanlop commented on src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:
```diff
@@ -397,7 +396,7 @@ private void CollectExtendedAttributesFromStandardFieldsIfNeeded()
 
             bool containsATime = _extendedAttributes.ContainsKey(PaxEaATime);
-            bool containsCTime = _extendedAttributes.ContainsKey(PaxEaATime);
+            bool containsCTime = _extendedAttributes.ContainsKey(PaxEaCTime);
```
We do, kinda:

- The `TarWriter_WriteEntry_Pax_Tests.WritePaxAttributes_Timestamps_AutomaticallyAdded` checks this for the `PaxTarEntry(entryFormat, string)` constructor, when the user does not explicitly set atime and ctime.
- The `PaxTarEntry_Conversion_Tests.Constructor_ConversionFrom*` tests are semi-related: they check that atime and ctime are always in the dictionary after conversion, in the constructor itself, not at write time.
- We don't have a test for the conversion constructors, so I'm adding one.

But now that you mention it, it seems these two conditions will never be false, making the code unreachable, so I am removing the code. Here's why:

When constructing a `PaxTarEntry`, whether it is new or it is being converted from another entry, all constructors ensure to add `atime` and `ctime` to the extended attributes dictionary if the user did not provide them (see the method `AddNewAccessAndChangeTimestampsIfNotExist`). Consider also that the extended attributes dictionary is exposed to the user as an `IReadOnlyDictionary<string, string>` getter-only property, so no new items can be added after construction.

carlossanlop committed:
TarHelpers methods can be private

carlossanlop committed:
Remove unreachable code for adding ctime and atime before writing Pax entry. Add tests to ensure we always add those entries to the dictionary on construction.

carlossanlop committed:
Fix typo in test when retrieving ctime.

carlossanlop committed:
Make sure CTime adds nanoseconds on Unix when retrieving info from disk.

carlossanlop commented:
I'm rebasing on top of the latest bits in main, which contain the fixes for the timestamp and Apple-specific test failures.

carlossanlop commented:
/azp run runtime-extra-platforms

eerhardt reviewed (approved):
Looks good! Thanks @carlossanlop.

carlossanlop commented:
The [tvOS failure](https://dev.azure.com/dnceng/public/_build/results?buildId=1840021&view=ms.vss-test-web.build-test-results-tab&runId=48593844&resultId=100888&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab) is unrelated to this change, and it is not critical:

The `ExtractToFile_SpecialFile_Unelevated_Throws` method threw `UnauthorizedAccessException` when attempting to extract the fifo from the archive into disk:

https://github.com/dotnet/runtime/blame/051b4828c7d3a0cad3289830ef9fd2120f45bb2b/src/libraries/System.Formats.Tar/tests/TarReader/TarReader.ExtractToFile.Tests.Unix.cs#L39

In other Unix platforms, extracting fifos does not throw, but apparently it does on tvOS. I'll get it fixed later.

Here's the callstack:
```
[14:55:07.7170830] 2022-06-22 14:55:07.706 System.Formats.Tar.Tests[58680:78698627]    Exception messages: System.UnauthorizedAccessException : Access to the path '/private/var/mobile/Containers/Data/Application/11374FD2-B51C-49C4-844D-F40F062ACE94/tmp/1pslf5q1.gf2/output' is denied.
[14:55:07.7171130] ---- System.IO.IOException : Operation not permitted
[14:55:07.7189880] 2022-06-22 14:55:07.708 System.Formats.Tar.Tests[58680:78698627]    Exception stack traces:    at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
[14:55:07.7190220]    at System.Formats.Tar.TarEntry.ExtractAsFifo(String destinationFileName)
[14:55:07.7190300] 2022-06-22 14:55:07.708 System.Formats.Tar.Tests[58680:78698627]    at System.Formats.Tar.TarEntry.ExtractToFileInternal(String filePath, String linkTargetPath, Boolean overwrite)
[14:55:07.7190360]    at System.Formats.Tar.TarEntry.ExtractToFile(String destinationFileName, Boolean overwrite)
[14:55:07.7201670] 2022-06-22 14:55:07.709 System.Formats.Tar.Tests[58680:78698627]    at System.Formats.Tar.Tests.TarReader_ExtractToFile_Tests.ExtractToFile_SpecialFile_Unelevated_Throws()
```
